### ISSUE TITLE
get_data (via load_niimg) casts to float64 (high memory usage)

### ISSUE THREAD

**Issue by adelavega** [nilearn/nilearn#2209]
What version of Nilearn are you using? master branch
What were you trying to do? Load a functional image (stored as integer type)
What did you expect will happen? Load in float32
What actually happened? It cast to float64 prior to casting back to float32

---

Due to https://github.com/nipy/nibabel/issues/832, nilearn's implementation of `get_data`, will always cast image data to `float64` even when dtype requested is `float32`.

This leads to really high, but temporary memory spikes. For example, loading a `.nii.gz` 4d image that is `int16` on disk, and ~398M:

```
img = load_niimg(img_path, dtype='float32')
```

Has the following memory profile:
![Figure_1](https://user-images.githubusercontent.com/2774448/68696165-78c33800-054a-11ea-9ba8-f292a7991d35.png)

A fix for this is coming in nibabel (https://github.com/nipy/nibabel/pull/833), but it will be using `get_fdata`, and in nibabel 3.0. Using this fix, I get the following memory profile for the same file:
![Figure_2](https://user-images.githubusercontent.com/2774448/68697195-57634b80-054c-11ea-803d-0979ade97006.png)

I understand that nilearn wishes to continue using `get_data` that will be deprecated in `nibabel` (https://github.com/nipy/nibabel/issues/834)

One potential compromise would be to continue using `get_data` (now coped to here) when the user does not specify a float type, but when a dtype=float is specified, use `nibabel.get_fdata`. This would ensure some backwards compatability while still allowing more memory efficient data loading.

**Comment by GaelVaroquaux** [comment]
> One potential compromise would be to continue using get_data (now coped to here) when the user does not specify a float type, but when a dtype=float is specified, use nibabel.get_fdata.

Sounds right!

Do you want to send a PR?

**Comment by adelavega** [comment]
Sure

**Comment by jeromedockes** [comment]
why not use `image.load_img(img).get_fdata()` ? or `nibabel.load_img(img).get_fdata()` ?

**Comment by adelavega** [comment]
That's exactly what I'm suggesting, but there seems to be interesting in maintaining previous behavior in `load_niimg` for backwards compatibility. So I'm suggesting only using `get_fdata` when a `dtype` is specified. 

I'm suggesting this change so that `nilearn` and other programs that use `load_niimg` will benefit from the memory improvements.

**Comment by effigies** [comment]
If nilearn *internally* uses `get_data()` and it doesn't take into account the target `dtype`, it will cast some data to float64 unnecessarily. This will have benefits even if no user ever touches nilearn's `get_data()`.

**Comment by jeromedockes** [comment]
If I understand correctly your PR will add a switch inside `nilearn._utils.niimg._get_data`?

**Comment by adelavega** [comment]
Yes, based on if a `dtype` is specified. Might not get to it until next year though.

**Comment by Remi-Gau** [comment]
probably worth checking if this problem still exists

**Comment by mathdugre** [comment]
take

**Comment by mathdugre** [comment]
It seems the issue is now solved, with nilearn's `load_niimg` and nibabel's `get_fdata` having the memory usage while loading a ~1GiB image.

# Memory usage
## Nilearn
```python
from pathlib import Path

from nilearn._utils.niimg import load_niimg

img_path = Path(__file__).parent / "rand_img.nii.gz"

img = load_niimg(img_path, dtype="float32")
```
<img width="1106" height="450" alt="Image" src="https://github.com/user-attachments/assets/50c71ea5-40dc-46af-a67e-a8694bf14baa" />

## Nibabel (get_fdata)
```python
from pathlib import Path

import nibabel as nib

img_path = Path(__file__).parent / "rand_img.nii.gz"

img = nib.load(img_path)
img.get_fdata(dtype="float32")

```
<img width="1106" height="450" alt="Image" src="https://github.com/user-attachments/assets/d3e3a7ce-b91d-4a8b-87ac-0e9b6aedefeb" />
